L1Trigger/TrackerTFP/src/KalmanFilter.cc

```diff
@@ -0,0 +1,398 @@
+    }
+  }
+
+  // fill output products
+  void KalmanFilter::produce(TTDTC::Streams& acceptedStubs, StreamsTrack& acceptedTracks, TTDTC::Streams& lostStubs, StreamsTrack& lostTracks) {
+    auto put = [this](const deque<State*>& states, TTDTC::Streams& streamsStubs, StreamsTrack& streamsTracks, int channel) {
+      const int streamId = region_ * dataFormats_->numChannel(Process::kf) + channel;
+      const int offset = streamId * setup_->numLayers();
+      StreamTrack& tracks = streamsTracks[streamId];
+      tracks.reserve(states.size());
+      for (int layer = 0; layer < setup_->numLayers(); layer++)
+        streamsStubs[offset + layer].reserve(states.size());
+      for (State* state : states) {
+        tracks.emplace_back(state->frame());
+        for (const StubKF& stub : state->stubs())
+          streamsStubs[offset + stub.layer()].emplace_back(stub.frame());
+        // adding a gap to all layer without a stub
+        for (int layer : state->hitPattern().ids(false))
+          streamsStubs[offset + layer].emplace_back(TTDTC::Frame());
+      }
+    };
+    for (int channel = 0; channel < dataFormats_->numChannel(Process::kf); channel++) {
+      deque<State*> stream;
+      deque<State*> lost;
+      // proto state creation
+      int trackId(0);
+      for (TrackKFin* track : input_[channel]) {
+        State* state = nullptr;
+        if (track) {
+          states_.emplace_back(dataFormats_, track, trackId++);
+          state = &states_.back();
+        }
+        stream.push_back(state);
+      }
+      // state propagation
+      for (layer_ = 0; layer_ < setup_->numLayers(); layer_++)
+        layer(stream);
+      // untruncated best state selection
+      deque<State*> untruncatedStream = stream;
+      accumulator(untruncatedStream);
+      // apply truncation
+      if (enableTruncation_ && (int)stream.size() > setup_->numFrames())
+      //if (enableTruncation_ && (int)stream.size() > setup_->numFramesIO())
+        stream.resize(setup_->numFrames());
+      // best state per candidate selection
+      accumulator(stream);
+      deque<State*> truncatedStream = stream;
+      // storing of best states missed due to truncation
+      sort(untruncatedStream.begin(), untruncatedStream.end());
+      sort(truncatedStream.begin(), truncatedStream.end());
+      set_difference(untruncatedStream.begin(), untruncatedStream.end(), truncatedStream.begin(), truncatedStream.end(), back_inserter(lost));
+      // store found tracks
+      put(stream, acceptedStubs, acceptedTracks, channel);
+      // store lost tracks
+      put(lost, lostStubs, lostTracks, channel);
+    }
+  }
+
+  // adds a layer to states
+  void KalmanFilter::layer(deque<State*>& stream) {
+    static constexpr int latency = 5;
+    // dynamic state container for clock accurate emulation
+    deque<State*> streamOutput;
+    deque<State*> stack;
+    // static delay container
+    vector<State*> delay(latency, nullptr);
+    delay.reserve(latency + 1);
```

Comment:
Given that the size of "delay" was declared on the previous line, and that this size is never changed, why do we need to call delay.reserve?

Reply:
to avoid unnecessary reallocation